R v D states.  
State Population  

## Import

In [1]:
import pandas as pd
import numpy as np

In [2]:
header = list(pd.read_csv('data/COVID-19_Case_Surveillance_Public_Use_Data_with_Geography.csv', nrows=1))

# one million times five
chunksize = 1000000 * 5

tfr = pd.read_csv('data/COVID-19_Case_Surveillance_Public_Use_Data_with_Geography.csv',
                  chunksize = chunksize,
                  iterator=True, dtype = {#'state_fips_code' : np.float32,
                                          #'county_fips_code' : np.float32, 
                                          'case_positive_specimen_interval' : np.float32, 
                                          'case_onset_interval' : np.float32
                                         },
                  usecols = [i for i in header if i != 'state_fips_code' or i != 'county_fips_code']
                 )

df = pd.concat(tfr, ignore_index=True)

df

C:\Users\Deric\anaconda3\envs\learn-env\lib\site-packages\IPython\core\interactiveshell.py:3343: DtypeWarning: Columns (18) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


case_month res_state  state_fips_code res_county  county_fips_code  \
0           2020-02       NaN              NaN        NaN               NaN   
1           2020-02       NaN              NaN        NaN               NaN   
2           2020-02       NaN              NaN        NaN               NaN   
3           2020-08       NaN              NaN        NaN               NaN   
4           2020-08       NaN              NaN        NaN               NaN   
...             ...       ...              ...        ...               ...   
22507134    2020-12        AZ              4.0       YUMA            4027.0   
22507135    2020-12        AZ              4.0       YUMA            4027.0   
22507136    2020-12        AZ              4.0       YUMA            4027.0   
22507137    2020-12        AZ              4.0       YUMA            4027.0   
22507138    2020-12        AZ              4.0       YUMA            4027.0   

          age_group   sex   race            ethnicity  \
0               NaN   NaN    NaN                  NaN   
1               NaN   NaN    NaN                  NaN   
2               NaN   NaN    NaN                  NaN   
3               NaN   NaN    NaN                  NaN   
4               NaN   NaN    NaN                  NaN   
...             ...   ...    ...                  ...   
22507134  65+ years  Male  White  Non-Hispanic/Latino   
22507135  65+ years  Male  White  Non-Hispanic/Latino   
22507136  65+ years  Male  White  Non-Hispanic/Latino   
22507137  65+ years  Male  White  Non-Hispanic/Latino   
22507138  65+ years  Male  White  Non-Hispanic/Latino   

          case_positive_specimen_interval  case_onset_interval  \
0                                     NaN                  0.0   
1                                     3.0                  0.0   
2                                     NaN                  0.0   
3                                     0.0                  NaN   
4                                     0.0                  NaN   
...                                   ...                  ...   
22507134                              NaN                  NaN   
22507135                              NaN                  NaN   
22507136                              NaN                  NaN   
22507137                              NaN                  NaN   
22507138                              NaN                  NaN   

                       process exposure_yn             current_status  \
0                      Missing     Missing  Laboratory-confirmed case   
1          Clinical evaluation         Yes  Laboratory-confirmed case   
2          Clinical evaluation     Missing  Laboratory-confirmed case   
3         Routine surveillance     Missing  Laboratory-confirmed case   
4         Routine surveillance     Missing  Laboratory-confirmed case   
...                        ...         ...                        ...   
22507134               Missing     Missing  Laboratory-confirmed case   
22507135               Missing     Missing  Laboratory-confirmed case   
22507136               Missing     Missing  Laboratory-confirmed case   
22507137               Missing     Missing  Laboratory-confirmed case   
22507138               Missing     Missing  Laboratory-confirmed case   

         symptom_status  hosp_yn   icu_yn death_yn underlying_conditions_yn  
0           Symptomatic      Yes  Missing      NaN                      NaN  
1           Symptomatic      Yes      Yes      NaN                      Yes  
2           Symptomatic      Yes       No      NaN                      Yes  
3          Asymptomatic       No       No  Missing                      Yes  
4               Missing  Missing  Missing  Missing                      Yes  
...                 ...      ...      ...      ...                      ...  
22507134        Missing      Yes  Missing      Yes                      NaN  
22507135        Missing  Missing  Missing      Yes                      NaN  
225

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22507139 entries, 0 to 22507138
Data columns (total 19 columns):
 #   Column                           Dtype  
---  ------                           -----  
 0   case_month                       object 
 1   res_state                        object 
 2   state_fips_code                  float64
 3   res_county                       object 
 4   county_fips_code                 float64
 5   age_group                        object 
 6   sex                              object 
 7   race                             object 
 8   ethnicity                        object 
 9   case_positive_specimen_interval  float32
 10  case_onset_interval              float32
 11  process                          object 
 12  exposure_yn                      object 
 13  current_status                   object 
 14  symptom_status                   object 
 15  hosp_yn                          object 
 16  icu_yn                           object 
 17  death_

## Cleaning

First thing first, I am going to delete the features that I don't think I want in my dataset. I am doing this first to help with the performance of the data set for the rest of the way  

I have both state and county names so I am dropping the state and county fips features  

At this moment I dont want the specimen interval features in the dataset.

In [4]:
df.drop(columns=['state_fips_code',
                 'county_fips_code',
                 'case_positive_specimen_interval',
                 'case_onset_interval'],
        inplace=True)

### res state

In [5]:
df_missing_state = df[df['res_state'].isna()]
df_missing_state.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 682 entries, 0 to 21789118
Data columns (total 15 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   case_month                682 non-null    object
 1   res_state                 0 non-null      object
 2   res_county                0 non-null      object
 3   age_group                 11 non-null     object
 4   sex                       0 non-null      object
 5   race                      0 non-null      object
 6   ethnicity                 0 non-null      object
 7   process                   682 non-null    object
 8   exposure_yn               682 non-null    object
 9   current_status            682 non-null    object
 10  symptom_status            682 non-null    object
 11  hosp_yn                   682 non-null    object
 12  icu_yn                    682 non-null    object
 13  death_yn                  609 non-null    object
 14  underlying_conditions

>anytime a person's 'state' is missing, we are also missing the person's 'county', 'sex', 'race', 'ethnicity', 'age group' and most are missing 'underlying condition yn'.

>for now, I am going reduce the dataset of the 682 entries to 22,506,457 

In [6]:
df['res_state'].describe()

count     22506457
unique          54
top             CA
freq       3547519
Name: res_state, dtype: object

In [7]:
df = df[df['res_state'].notnull()]

> Res states include all 50 state in the United State plus these four extras:  
> PR - Puerto Rico  
> DC - Washington DC  
> GU - Guam  
> VI - Virgin Islands

### res_county

In [8]:
df['res_county'].isna().sum()

1614807

In [9]:
df[df['res_county'].isna()]['res_state'].value_counts()

IN    277061
GA    139212
IA     94077
TN     65958
KY     60671
KS     59604
IL     59511
VA     57901
AL     51862
AR     46718
OK     43972
NE     43341
DC     42644
MN     40566
WI     39620
MS     32665
SD     32550
TX     30434
MT     30150
UT     26587
CO     24889
NC     24052
PR     23988
MD     19415
ID     19371
MI     19333
FL     19292
MO     17307
NM     14604
LA     13107
OH     12873
AK     10826
SC     10201
NY      9721
NJ      9629
WA      9307
PA      8259
WV      7798
GU      7067
ND      6914
CA      6871
OR      6532
ME      6345
VT      5566
NV      4884
MA      4074
AZ      3665
HI      2951
DE      2897
VI      2588
NH      2294
CT      1450
WY      1276
RI       357
Name: res_state, dtype: int64

In [10]:
len(df[df['res_county'].isna()]['res_state'].value_counts())

54

> 'res_county' feature is missing 1,614,807 entries spread out between all 54 'res_state's

In [11]:
df['res_county'].fillna('Missing', inplace=True)

C:\Users\Deric\anaconda3\envs\learn-env\lib\site-packages\pandas\core\series.py:4536: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  downcast=downcast,


### case_month

In [12]:
df['case_month'].isna().sum()

0

In [13]:
df['case_month'].describe()

count     22506457
unique          15
top        2020-12
freq       4852575
Name: case_month, dtype: object

In [14]:
df['case_month'].value_counts()

2020-12    4852575
2021-01    3970329
2020-11    3482415
2020-10    1757756
2021-02    1568200
2020-07    1376516
2020-09    1047257
2020-08    1030610
2020-06     880507
2020-04     783269
2021-03     714640
2020-05     667374
2020-03     367284
2020-02       4589
2020-01       3136
Name: case_month, dtype: int64

Typically I would separate the year and the month, However there is only 15 months and over 22 million data entries, I am going to leave it in one column for now.

### age_group

In [15]:
df['age_group'].isna().sum()

242099

In [16]:
df['age_group'].fillna('Missing', inplace=True)

C:\Users\Deric\anaconda3\envs\learn-env\lib\site-packages\pandas\core\series.py:4536: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  downcast=downcast,


In [17]:
df['age_group'].value_counts()

18 to 49 years    11911259
50 to 64 years     4525363
65+ years          3109385
0 - 17 years       2559165
Missing             401285
Name: age_group, dtype: int64

### sex

In [18]:
df['sex'].isna().sum()

644342

In [19]:
df['age_group'].fillna('Missing', inplace=True)

In [20]:
df['sex'].value_counts()

Female     11337841
Male       10334629
Unknown      158104
Missing       31530
Other            11
Name: sex, dtype: int64

### race

In [21]:
df['race'].isna().sum()

3753372

In [22]:
df['race'].fillna('Missing', inplace=True)

In [23]:
df['race'].value_counts()

White                                     9540953
Missing                                   5714493
Unknown                                   3712716
Black                                     1670933
Multiple/Other                            1227259
Asian                                      459550
American Indian/Alaska Native              152382
Native Hawaiian/Other Pacific Islander      28171
Name: race, dtype: int64

### ethnicity

In [24]:
df['ethnicity'].value_counts()

Non-Hispanic/Latino    9168555
Unknown                5694524
Missing                1664232
Hispanic/Latino        1487413
Name: ethnicity, dtype: int64

>I understand that there is a subtle difference between race and ethnicity, Since I have a more detailed feature of race and the feature ethnicity dosen't account for many categories, I am going to delete this feature and just rely on race.

In [25]:
df.drop(columns=['ethnicity'], inplace=True)

C:\Users\Deric\anaconda3\envs\learn-env\lib\site-packages\pandas\core\frame.py:4174: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


### process

In [26]:
df['process'].isna().sum()

0

In [27]:
df['process'].value_counts()

Missing                            21543962
Clinical evaluation                  537966
Routine surveillance                 207188
Multiple                             115946
Contact tracing of case patient       70068
Unknown                               22030
Other                                  9297
Name: process, dtype: int64

### exposure_yn

In [28]:
df['exposure_yn'].isna().sum()

0

In [29]:
df['exposure_yn'].value_counts()

Missing    20434459
Yes         1551517
Unknown      520481
Name: exposure_yn, dtype: int64

### symptom_status

In [30]:
df['symptom_status'].isna().sum()

0

In [31]:
df['symptom_status'].value_counts()

Symptomatic     9696774
Missing         8316489
Unknown         4109737
Asymptomatic     383457
Name: symptom_status, dtype: int64

### hosp_yn

In [32]:
df['hosp_yn'].isna().sum()

0

In [33]:
df['hosp_yn'].value_counts()

Missing    9882363
No         8031678
Unknown    3513930
Yes        1078486
Name: hosp_yn, dtype: int64

### icu_yn

In [34]:
df['icu_yn'].isna().sum()

0

In [35]:
df['icu_yn'].value_counts()

Missing    18402205
Unknown     2690360
No          1311728
Yes          102164
Name: icu_yn, dtype: int64

### death_yn

In [36]:
df['death_yn'].isna().sum()

1043414

In [37]:
df['death_yn'].fillna('Missing', inplace=True)

In [38]:
df['death_yn'].value_counts()

No         11066699
Missing     9034867
Unknown     2159863
Yes          245028
Name: death_yn, dtype: int64

### underlying_conditions_yn

In [39]:
df['underlying_conditions_yn'].isna().sum()

21006836

In [40]:
df['underlying_conditions_yn'].fillna('Missing', inplace=True)

In [41]:
df['underlying_conditions_yn'].value_counts()

Missing    21006836
Yes         1482254
No            17367
Name: underlying_conditions_yn, dtype: int64

## Export

Due to how large the dataset is, I'm going to continue the EDA in a separate notebook.

In [42]:
df

case_month res_state res_county  age_group   sex     race  \
66          2020-04        AK    Missing    Missing   NaN  Missing   
67          2020-07        AL    Missing    Missing   NaN  Missing   
68          2020-07        AL    Missing    Missing   NaN  Missing   
69          2021-02        AL    Missing    Missing   NaN  Missing   
70          2021-02        AL    Missing    Missing   NaN  Missing   
...             ...       ...        ...        ...   ...      ...   
22507134    2020-12        AZ       YUMA  65+ years  Male    White   
22507135    2020-12        AZ       YUMA  65+ years  Male    White   
22507136    2020-12        AZ       YUMA  65+ years  Male    White   
22507137    2020-12        AZ       YUMA  65+ years  Male    White   
22507138    2020-12        AZ       YUMA  65+ years  Male    White   

                       process exposure_yn             current_status  \
66        Routine surveillance         Yes  Laboratory-confirmed case   
67                     Missing     Missing  Laboratory-confirmed case   
68                     Missing     Missing  Laboratory-confirmed case   
69                     Missing     Missing  Laboratory-confirmed case   
70                     Missing     Missing  Laboratory-confirmed case   
...                        ...         ...                        ...   
22507134               Missing     Missing  Laboratory-confirmed case   
22507135               Missing     Missing  Laboratory-confirmed case   
22507136               Missing     Missing  Laboratory-confirmed case   
22507137               Missing     Missing  Laboratory-confirmed case   
22507138               Missing     Missing  Laboratory-confirmed case   

         symptom_status  hosp_yn   icu_yn death_yn underlying_conditions_yn  
66          Symptomatic      Yes       No  Missing                      Yes  
67          Symptomatic      Yes  Missing  Missing                  Missing  
68          Symptomatic      Yes  Missing  Missing                  Missing  
69              Missing  Missing  Missing  Missing                  Missing  
70              Missing  Missing  Missing  Missing                  Missing  
...                 ...      ...      ...      ...                      ...  
22507134        Missing      Yes  Missing      Yes                  Missing  
22507135        Missing  Missing  Missing      Yes                  Missing  
22507136        Missing      Yes  Missing      Yes                  Missing  
22507137        Missing      Yes  Missing      Yes                  Missing  
22507138        Missing      Yes  Missing      Yes                  Missing  

[22506457 rows x 14 columns]

In [43]:
df.to_csv('data/cleaned_cdc_covid19.csv')